src/backends/torch/optim/ranger.cc

```diff
@@ -0,0 +1,229 @@
+            auto &options = static_cast<RangerOptions &>(group.options());
+
+            // State initialization
+            if (param_state == state_.end())
+              {
+                auto state = std::make_unique<RangerParamState>();
+                state->step(0);
+                // Exponential moving average of gradient values
+                state->exp_avg(
+                    torch::zeros_like(p, torch::MemoryFormat::Preserve));
+                // Exponential moving average of squared gradient values
+                state->exp_avg_sq(
+                    torch::zeros_like(p, torch::MemoryFormat::Preserve));
+                state->slow_buffer(
+                    torch::empty_like(p, torch::MemoryFormat::Preserve));
+                state->slow_buffer().copy_(p.data());
+                state_[c10::guts::to_string(p.unsafeGetTensorImpl())]
+                    = std::move(state);
+              }
+
+            auto &state = static_cast<RangerParamState &>(
+                *state_[c10::guts::to_string(p.unsafeGetTensorImpl())]);
+            auto &exp_avg = state.exp_avg();
+            auto &exp_avg_sq = state.exp_avg_sq();
+
+            state.step(state.step() + 1);
+            auto beta1 = std::get<0>(options.betas());
+            auto beta2 = std::get<1>(options.betas());
+
+            auto bias_correction1 = 1.0 - std::pow(beta1, state.step());
+            auto bias_correction2 = 1.0 - std::pow(beta2, state.step());
+
+            if (options.weight_decay() != 0) // weight decay not decoupled !!
+              grad = grad.add(p, options.weight_decay());
+
+            if (options.gradient_centralization())
+              {
+
+                std::vector<long int> dim;
+                for (long int i = 1; i < grad.dim(); ++i)
+                  dim.push_back(i);
+                grad.add_(-grad.mean(torch::IntArrayRef(dim), true));
+              }
+
+            exp_avg.mul_(beta1).add_(grad, 1 - beta1); // m_t
+
+            if (options.adabelief())
+              exp_avg_sq.mul_(beta2).addcmul_(grad - exp_avg, grad - exp_avg,
+                                              1 - beta2); // v_t
+            else
+              exp_avg_sq.mul_(beta2).addcmul_(grad, grad, 1 - beta2); // v_t
+            torch::Tensor denom;
+            if (!options.rectified())
+              {
+                denom = (exp_avg_sq.sqrt() / sqrt(bias_correction2))
+                            .add_(options.eps());
+
+                auto step_size = options.lr() / bias_correction1;
+                p.addcdiv_(exp_avg, denom, -step_size);
+              }
+            else
+              {
+                auto N_sma_max = 2.0 / (1.0 - beta2) - 1.0; // rho_inf
+                auto beta2_t = std::pow(beta2, state.step());
+                auto N_sma
+                    = N_sma_max
+                      - 2.0 * (float)state.step() * beta2_t / bias_correction2;
+                double step_size;
+                if (N_sma >= 5)
```

Comment:
FYI this value appears to differ from the one that you use with Caffe https://github.com/jolibrain/caffe/blob/6c9b6bee92a2825d50be07c8a9e523ed56bfef20/src/caffe/solvers/adam_solver.cpp#L110

Reply:
not really , this one states `>=5`, while the other states `>4` and N_sma is supposed to be an int